proxy_http.go

```diff
@@ -0,0 +1,302 @@
+		onRequest:           onRequest,
+		onResponse:          onResponse,
+		onError:             onError,
+		dial:                dial,
+	}
+	return ic.intercept
+}
+
+type httpInterceptor struct {
+	discardFirstRequest bool
+	idleTimeout         time.Duration
+	onRequest           func(req *http.Request) *http.Request
+	onResponse          func(resp *http.Response) *http.Response
+	onError             func(req *http.Request, err error) *http.Response
+	dial                DialFunc
+}
+
+func (ic *httpInterceptor) intercept(w http.ResponseWriter, req *http.Request) error {
+	var downstream net.Conn
+	var downstreamBuffered *bufio.ReadWriter
+	tr := &http.Transport{
+		Dial:                ic.dial,
+		IdleConnTimeout:     ic.idleTimeout,
+		MaxIdleConnsPerHost: 8, // just to keep clients from getting too greedy
+	}
+	var err error
+
+	closeDownstream := false
+	defer func() {
+		if closeDownstream {
+			if closeErr := downstream.Close(); closeErr != nil {
+				log.Tracef("Error closing downstream connection: %s", closeErr)
+			}
+		}
+		tr.CloseIdleConnections()
+	}()
+
+	// Hijack underlying connection.
+	downstream, downstreamBuffered, err = w.(http.Hijacker).Hijack()
+	if err != nil {
+		fullErr := errors.New("Unable to hijack connection: %s", err)
+		respondBadGateway(w, fullErr)
+		return fullErr
+	}
+	closeDownstream = true
+
+	return ic.processRequests(req.RemoteAddr, req, downstream, downstreamBuffered, tr)
+}
+
+func (ic *httpInterceptor) processRequests(remoteAddr string, req *http.Request, downstream net.Conn, downstreamBuffered *bufio.ReadWriter, tr *http.Transport) error {
+	var readErr error
+
+	first := true
+	for {
+		discardRequest := first && ic.discardFirstRequest
+		if discardRequest {
+			err := ic.onRequest(req).Write(ioutil.Discard)
+			if err != nil {
+				return errors.New("Error discarding first request: %v", err)
+			}
+		} else {
+			resp, err := tr.RoundTrip(prepareRequest(ic.onRequest(req)))
+			if err != nil {
+				errResp := ic.onError(req, err)
+				if errResp != nil {
+					errResp.Request = req
+					ic.writeResponse(downstream, errResp)
+				}
+				return err
+			}
+			writeErr := ic.writeResponse(downstream, resp)
+			if writeErr != nil {
+				if isUnexpected(writeErr) {
+					return errors.New("Unable to write response to downstream: %v", writeErr)
+				}
+				// Error is not unexpected, but we're done
+				return nil
+			}
+		}
+
+		if req.Close {
+			// Client signaled that they would close the connection after this
+			// request, finish
+			return nil
+		}
+
+		req, readErr = http.ReadRequest(downstreamBuffered.Reader)
+		if readErr != nil {
+			if isUnexpected(readErr) {
+				return errors.New("Unable to read next request from downstream: %v", readErr)
+			}
+			return nil
+		}
+
+		// Preserve remote address from original request
+		req.RemoteAddr = remoteAddr
+		first = false
+	}
+}
+
+func (ic *httpInterceptor) writeResponse(downstream net.Conn, resp *http.Response) error {
```

Comment:
Changed type to `io.Writer`, kept name of `downstream` because `w` is used elsewhere to refer to the `ResponseWriter`.